pyzim/pyzim.pyx

```diff
@@ -1,338 +1,183 @@
+cimport pyzim
+cimport cpython.ref as cpy_ref
+from cython.operator import dereference
+
+from libc.stdint cimport uint32_t, uint64_t
 from libcpp.string cimport string
 from libcpp cimport bool
 from libcpp.memory cimport shared_ptr, make_shared
 
 import datetime
 from collections import defaultdict
-from cython.operator import dereference
-
-cimport zim_wrapper as zim
-
-
 
 #########################
 #       ZimArticle      #
 #########################
 
+cdef class ZimBlob:
+    cdef Blob* c_blob
 
-cdef class ZimArticle:
-    """ 
-    A class to represent a Zim Article. 
-    
-    Attributes
-    ----------
-    *c_zim_article : zim.ZimArticle
-        a pointer to the C++ article object
-    _can_write : bool
-        flag if the article is ready for writing
-
-    Properties
-    -----------
-    namespace: str
-        the article namespace
-    title : str
-        the article title
-    content : str
-        the article content
-    longurl : str
-        the article long url i.e {NAMESPACE}/{redirect_url}
-    url : str
-        the article url
-    mimetype : str
-        the article mimetype
-    is_redirect : bool
-        flag if the article is a redirect 
-    redirect_longurl: str
-        the long redirect url i.e {NAMESPACE}/{redirect_url}
-    redirect_url : str
-        the article url
-
-    Methods
-    -------
-
-    gamma(n=1.0)
-        Change the photo's gamma exposure.
-
-    """
-    cdef zim.ZimArticle *c_zim_article
-    cdef object _can_write
-
-    VALID_NAMESPACES = ["-","A","B","I","J","M","U","V","W","X"]
-
-
-    def __cinit__(self, url="", str content="", namespace= "A", mimetype= "text/html", title="", redirect_article_url= "",filename="", should_index=True ):
-
-        # Encoding must be set to UTF-8 
-        #cdef bytes py_bytes = content.encode(encoding='UTF-8')
-        #cdef char* c_string = py_bytes
-        
-        if namespace not in self.VALID_NAMESPACES:
-            raise RuntimeError("Invalid Namespace")
-
-        c_zim_art = new zim.ZimArticle(ord(namespace),                     # Namespace
-                                       url.encode('UTF-8'),                # url
-                                       title.encode('UTF-8'),              # title
-                                       mimetype.encode('UTF-8'),           # mimeType
-                                       redirect_article_url.encode('UTF-8'),# redirectUrl
-                                       should_index,                                # shouldIndex
-                                       content.encode('UTF-8'))
-        self.__setup(c_zim_art)
+    def __init__(self, bytes content):
+        self.c_blob = new Blob(<char *> content, len(content))
```

Comment:
> I'm not sure to understand the status of this PR.

This is a WIP using the cython blend approach. Currently I am addressing issues from the first version #2 . That will help me a lot defining the API and understanding the use cases. The idea es to keep the same API and other than using the article subclasses for each task.

> What we discuss about is to have two different wrapper for the articles but I see only one.

Locally it's more advanced (unpushed) . I do have a read article as follows:

```python3
cdef class ZimReadArticle(ZimArticle):
    cdef Article* c_read_article

    def __init__(self):
        ZimArticle.__init__(self)

    cdef __setup(self, Article *art):
        """Assigns an internal pointer to the wrapped C++ article object.
        
        Parameters
        ----------
        *art : zim.ZimArticle
            Pointer to a C++ article object
        """

        if self.c_read_article != NULL:
            del self.c_read_article

        # Set new internal C zim.ZimArticle article
        self.c_read_article = art

    @staticmethod
    cdef from_read_article(Article art):
        """Creates a python ZimReadArticle from a C++ zim Article article.
        
        Parameters
        ----------
        art : Article
            A C++ zim Article read with File

        Return
        ------
        article: ZimReadArticle
            A initialiazed ZimReadArticle
        """
        cdef ZimReadArticle article = ZimReadArticle()
        article.__setup(<Article *> &art)
        return article

    def is_redirect(self):
       # call is_redirect
        pass


    @property
    def can_write(self):
        return True

    def get_url(self):
        pass

    def get_title(self):
        return self.c_read_article.getTitle().decode('UTF-8')
    
    def get_mime_type(self):
        return  self.c_read_article.getMimeType().decode('UTF-8')
    
    def get_filename(self):
        return ""
  

    def get_data(self):
        return self.c_read_article.getTitle()
        c_blob = self.c_read_article.getData()
        return  ZimBlob.from_c_blob(&c_blob)
```

> In fact, it seems that there is nothing related to reading a zim File even if the `examples.py` use a `pyzim.ZimReader`.

Unpushed WIP.  The goal is to be able to use the same API regardless of the article type.

> 
> If we assume that this branch contains only things related to writing a zim :
> 
> * Please move all this code in a submodule (`write`/`writer`/`writing`/...).
> * Rename `Article`, `ZimArticle` and `ZimArticleWrapper` to something including the fact they are used to write zim file. (If may not be necessary if you move them in a specific submodule)
> 
> Let's be sure that we understand well on what is the separation between `zim::*` and `zim::writer::*` :
> They are two different libraries. They are in the same repository and compiled in the same `.so` but they have almost nothing in common.
> They could be separated in two different repositories without a lot of change (I think that they share the blob structure and one internal small header class only). This is not plan at all, but you should keep this in mind. Especially this is almost two different bindings to create.

This is understood, but the idea I had at first was a pythonic api that doesn't need to know internal libzim details. If you agree we can totally separate reading from writing.

This will however restrict the property uses, for example

python_content = "hola hola"
read_article.content = python_content

I will have to replace the content, currently obtained from the article, with a python constructed Blob. Any ideas on this?

Reply:
I'm not sure of what you call the "blend" approach.
If you means that there is only one python article to wrap the two cpp articles, I disagree with this approach.

> The goal is to be able to use the same API regardless of the article type.

That is the main point. We need different API because it is a different use case.

> If you agree we can totally separate reading from writing

I totally agree, this is what I want. This should be handled differently and separately.


> ```
> python_content = "hola hola"
> read_article.content = python_content
> ```

Either I don't understand your question or you don't understand the use case :)
A "read" article (zim::Article) is read only because it is a "view" of what is inside the zim file. You cannot change it's value. So you cannot change it's value using property (or anything else).
On the opposite side a "write" article is not a article but an interface the user code have to implement. So, no property neither.

Reply:
I'm curious to hear more about your reasoning for keeping the reader and writer separate in python @mgautierfr. To me it seems like most python users would intuitively try to use the same interface, e.g. a typical housekeeping process might be rewriting a few broken image URLs in a ZIM collection:

```python
from libzim import ZimArticle

...
def rewrite_img_urls(zim: ZimArticle):
    article_body = zim.content().decode('UTF-8')
    new_article_body = re.replace(article_body, ... some regex pattern ...)
    zim.content = new_article_body

for article in load_articles(...):
    rewrite_img_urls(article)
```